### ISSUE TITLE
memory leak occurs when creating a new type

### ISSUE THREAD

**Issue by zaxrok** [marshmallow-code/marshmallow#732]
When I use the following code,  I observe that memory leak is very fast 
**MsgSchema = type('MsgSchema', (marshmallow.Schema,), {})**
```
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import marshmallow
import time
import psutil
import os

if __name__ == '__main__':
    pid = os.getpid()
    process = psutil.Process(pid)
    while True:
        MsgSchema = type('MsgSchema', (marshmallow.Schema,), {})
        mem = process.get_memory_info()[0]
        print("memory usage(%s):%d" % (pid, mem))
        time.sleep(0.01)
```

**Comment by sloria** [comment]
Ah, this is because creating a new `Schema` class adds it to a [registry](https://github.com/marshmallow-code/marshmallow/blob/dev/marshmallow/class_registry.py) in memory. You can prevent dynamically-generated classes from getting entered into the registry by passing a blank name.

```python
 MsgSchema = type('', (marshmallow.Schema,), {})
```

Now that you mention this, though, I realize that we should only add a class to the registry if an existing registry with the same fully-qualified module path doesn't already exist.

**Comment by asmodehn** [comment]
The schema name is always the same here, so I would assume that the registry contains only one element, and that the other classes are deleted when needed. Is that a correct assumption ? Or is there anything preventing the memory to be freed in the schema class ?

**Comment by sloria** [comment]
Looks like we always register the fully-qualified module path, even if it's already in the registry

https://github.com/marshmallow-code/marshmallow/blob/642d18b58d9b42d16b450a30b42aa57ea3859192/marshmallow/class_registry.py#L55

This appears to be a bug. I'd gladly review and merge a PR fixing this.

**Comment by asmodehn** [comment]
Just to be clear, are you saying that there is **no need** for one classname **to hold a list** of classes, and instead, we should have **only one class per name in the registry**, and that a registered class entry gets **overwritten** if a new class is registered with the same name ?

**Comment by sloria** [comment]
No, we still need to store a list of classes for each name.

However, we don't need to store duplicate values in the lists.

**Comment by asmodehn** [comment]
> No, we still need to store a list of classes for each name.
However, we don't need to store duplicate values in the lists.

When storing serializer class in the registry, the module name is used to disambiguate between _same_ classes, where _same_ is defined by the fullpath (module+classname) [there](https://github.com/marshmallow-code/marshmallow/blob/dev/marshmallow/class_registry.py#L48).

However when dealing with fullpath [just under](https://github.com/marshmallow-code/marshmallow/blob/dev/marshmallow/class_registry.py#L55), that would mean that each new entry is understood as _same_ as an existing one. Therefore we don't need a list there anymore... Or we do need a way to differentiate between different classes that does not rely on their fullpath (which might not be suitable for complexity cost).

**Comment by asmodehn** [comment]
Found another related issue where storing a class with same name in the registry would erase classes with same name (but from different modules) from the list.

**Comment by sloria** [comment]
This is fixed in 2.16.3 and 3.0.0b20.
